design/src/rfcs/rfc0023_sdk_credential_cache_type_safety.md

```diff
@@ -0,0 +1,328 @@
+RFC: SDK Credential Cache Type Safety
+=====================================
+
+> Status: RFC
+>
+> Applies to: AWS SDK for Rust
+
+At time of writing (2022-10-11), the SDK's credentials provider can be customized by providing:
+
+1. A profile credentials file to modify the default provider chain
+2. An instance of one of the credentials providers implemented in `aws-config`, such as
+   the `AssumeRoleCredentialsProvider`, `ImdsCredentialsProvider`, and so on.
+3. A custom struct that implements the `ProvideCredentials`
+
+The problem this RFC examines is that when options 2 and 3 above are exercised, the customer
+needs to be aware of credentials caching and put additional effort to ensure caching is set up
+correctly (and that double caching doesn't occur). This is especially difficult to get right
+since some built-in credentials providers (such as `AssumeRoleCredentialsProvider`) already
+have caching, while most others do not and need to be wrapped in `LazyCachingCredentialsProvider`.
+
+The goal of this RFC is to create an API where Rust's type system ensures caching is set up
+correctly, or explicitly opted out of.
+
+`CredentialsCache` and `ConfigLoader::credentials_cache`
+--------------------------------------------------------
+
+A new config method named `credentials_cache()` will be added to `ConfigLoader` and the
+generated service `Config` builders that takes a `CredentialsCache` instance. This `CredentialsCache`
+will be a struct with several functions on it to create and configure the cache.
+
+Client creation will ultimately be responsible for taking this `CredentialsCache` instance
+and wrapping the given (or default) credentials provider.
+
+The `CredentialsCache` would look as follows:
+
+```rust
+enum Inner {
+    Lazy(LazyConfig),
+    // Eager doesn't exist today, so this is purely for illustration
+    Eager(EagerConfig),
+    // Custom may not be implemented right away
+    // Not naming or specifying the custom cache trait for now since its out of scope
+    Custom(Box<dyn SomeCacheTrait>),
+
+    NoCaching,
+}
+pub struct CredentialsCache {
+    inner: Inner,
+}
+
+impl CredentialsCache {
+    // Methods prefixed with `default_` just use the default cache settings
+    pub fn default_lazy() -> Self { /* ... */ }
+    pub fn default_eager() -> Self { /* ... */ }
```

Comment:
Should we specify that these shouldn't change implementation based on features? Currently, I think several things depend on their being an HTTP client and I know we have some features that enable that client.

We already have the features RFC, so maybe mentioning this would just be redundant.

Reply:
I think we're probably OK here given the other RFC.